### PR TITLE
Make remote-run use tar to transfer files.

### PR DESCRIPTION
**Pull request by al45tair**
Rather than using `sftp`, which seems to sometimes fail to transfer file mtimes properly on some of our test platforms, use `tar` to transfer files to the test system.

rdar://88179140

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
Looks like I need to fiddle with the `remote-run` tests as well.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by mikeash** [comment]
Possibly silly question, does the existing `sftp` code pass `-p` to preserve times and modes? I don't see it but I might have missed it.

On another note, it would be really cool if this could use `rsync` to speed up repeated transfers, but that's probably something for another day.

File: utils/remote-run [hunk 1]
```diff
@@ -39,17 +44,48 @@ class CommandRunner(object):
...
+                                dirinfo = archive.getmember(dirname)
+                            except KeyError:
+                                dirinfo = tarfile.TarInfo(posixpath.dirname(remote_file))
+                                dirinfo.type = tarfile.DIRTYPE
+                                dirinfo.mode = 0o755
+                                archive.addfile(dirinfo)
+                        else:
+                            raise
+
+            invocation = self.remote_invocation(['/usr/bin/tar', '-x', '-j', '-P',
```
Comment (compnerd) [inline utils/remote-run:73]: I tend to prefer to write this as:

`target jxf - -P -C /` as `j`, `x`, `f` are the original Unix arguments and do not actually require the `-` as per Unix style argument passing.

**Review by compnerd — APPROVED** [review]

**Comment by al45tair** [comment]
> Possibly silly question, does the existing `sftp` code pass `-p` to preserve times and modes? I don't see it but I might have missed it.

Not a silly question at all. I don't think it does, but I thought I'd tried `-p` and it didn't fix the problem. However, now you've mentioned it again I'm no longer 100% certain. I'll go and try `-p` again and if that fixes it then that's a better solution.
 
> On another note, it would be really cool if this could use `rsync` to speed up repeated transfers, but that's probably something for another day.

I thought about doing it with `rsync`. The problem is that we don't put the remote files in the same places we collect them from — even relative to one another — so we'd have to build the remote structure we want first before we could `rsync` it.

**Comment by al45tair** [comment]
So it turns out that `-p` definitely doesn't work. I did think I'd tested it and determined that already, but I hadn't noted it down anywhere and after the weekend I wasn't entirely positive that I'd got it right.

**Review by mikeash — APPROVED** [review]

**Comment by al45tair** [comment]
I want to see what people think, but I re-did this to use `rsync` instead and I think the result is somewhat better. See https://github.com/apple/swift/pull/41125

**Comment by al45tair** [comment]
I'm going to drop this one in favour of the rsync one.

**Pull request closed without merging**